src/System.Memory/tests/Memory/Trim.cs

```diff
@@ -0,0 +1,427 @@
+        [InlineData(new int[] { 1 }, new int[0], new int[] { 1 })]
+        [InlineData(new int[] { 1 }, new int[] { 1 }, new int[0])]
+        [InlineData(new int[] { 2 }, new int[] { 1 }, new int[] { 2 })]
+        [InlineData(new int[] { 1, 2, 1 }, new int[] { 1 }, new int[] { 1, 2 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 1 }, new int[] { 1, 2 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 2 }, new int[] { 1, 2, 1, 1 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 3 }, new int[] { 1, 2, 1, 1 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 1, 2 }, new int[0])]
+        [InlineData(new int[] { 3, 2, 1, 1 }, new int[] { 1, 2 }, new int[] { 3 })]
+        [InlineData(new int[] { 3, 2, 1, 1 }, new int[] { 1, 2, 4 }, new int[] { 3 })]
+        [InlineData(new int[] { 2, 1, 1, 1 }, new int[] { 1 }, new int[] { 2 })]
+        [InlineData(new int[] { 1, 1, 1, 1 }, new int[] { 1 }, new int[0])]
+        public static void MemoryExtensions_TrimEnd_Multi(int[] values, int[] trims, int[] expected)
+        {
+            Memory<int> memory = new Memory<int>(values).TrimEnd(trims);
+            Assert.True(System.Linq.Enumerable.SequenceEqual(expected, memory.ToArray()));
+
+            ReadOnlyMemory<int> rom = new ReadOnlyMemory<int>(values).TrimEnd(trims);
+            Assert.True(System.Linq.Enumerable.SequenceEqual(expected, rom.ToArray()));
+
+            Span<int> span = new Span<int>(values).TrimEnd(trims);
+            Assert.True(System.Linq.Enumerable.SequenceEqual(expected, span.ToArray()));
+
+            ReadOnlySpan<int> ros = new ReadOnlySpan<int>(values).TrimEnd(trims);
+            Assert.True(System.Linq.Enumerable.SequenceEqual(expected, ros.ToArray()));
+        }
+
+        [Theory]
+        [InlineData(new int[0], new int[0], new int[0])]
+        [InlineData(new int[0], new int[] { 1 }, new int[0])]
+        [InlineData(new int[] { 1 }, new int[0], new int[] { 1 })]
+        [InlineData(new int[] { 1 }, new int[] { 1 }, new int[0])]
+        [InlineData(new int[] { 2 }, new int[] { 1 }, new int[] { 2 })]
+        [InlineData(new int[] { 1, 2, 1 }, new int[] { 1 }, new int[] { 2 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 1 }, new int[] { 2 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 2 }, new int[] { 1, 2, 1, 1 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 3 }, new int[] { 1, 2, 1, 1 })]
+        [InlineData(new int[] { 1, 2, 1, 1 }, new int[] { 1, 2 }, new int[0])]
+        [InlineData(new int[] { 2, 1, 3, 2, 1, 1 }, new int[] { 1, 2 }, new int[] { 3 })]
+        [InlineData(new int[] { 2, 1, 3, 2, 1, 1 }, new int[] { 1, 2, 4 }, new int[] { 3 })]
+        [InlineData(new int[] { 1, 2, 1, 1, 1 }, new int[] { 1 }, new int[] { 2 })]
+        [InlineData(new int[] { 1, 1, 1, 1 }, new int[] { 1 }, new int[0])]
+        public static void MemoryExtensions_Trim_Multi(int[] values, int[] trims, int[] expected)
```

Comment:
Several of the tests are failing:
System.MemoryTests.MemoryTests/MemoryExtensions_Trim_Multi(values: [1, 1, 1, 1], trims: [1], expected: [])
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F35576~2Fmerge/test~2Ffunctional~2Fcli~2F/20190404.27/workItem/System.Memory.Tests/analysis/xunit/System.MemoryTests.MemoryTests~2FMemoryExtensions_Trim_Multi(values:%20%5B1,%201,%201,%201%5D,%20trims:%20%5B1%5D,%20expected:%20%5B%5D)
```text
Message :
System.ArgumentOutOfRangeException : Specified argument was out of the range of valid values.
Stack Trace :
   at System.Memory`1.Slice(Int32 start, Int32 length) in F:\vsagent\44\s\src\System.Private.CoreLib\shared\System\Memory.cs:line 273
   at System.MemoryExtensions.Trim[T](Memory`1 memory, T trimElement) in F:\vsagent\44\s\src\System.Private.CoreLib\shared\System\MemoryExtensions.Trim.cs:line 22
   at System.MemoryExtensions.Trim[T](Memory`1 memory, ReadOnlySpan`1 trimElements) in F:\vsagent\44\s\src\System.Private.CoreLib\shared\System\MemoryExtensions.Trim.cs:line 229
   at System.MemoryTests.MemoryTests.MemoryExtensions_Trim_Multi(Int32[] values, Int32[] trims, Int32[] expected) in D:\a\1\s\src\System.Memory\tests\Memory\Trim.cs:line 166
```

Also, System.Runtime.Tests:
 System.Tests.StringTests/MakeSureNoTrimChecksGoOutOfRange_Memory
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F35576~2Fmerge/test~2Ffunctional~2Fcli~2F/20190404.27/workItem/System.Runtime.Tests/analysis/xunit/System.Tests.StringTests~2FMakeSureNoTrimChecksGoOutOfRange_Memory
```text
Message :
Assert.True() Failure
Expected: True
Actual:   False
Stack Trace :
   at System.Tests.StringTests.MakeSureNoTrimChecksGoOutOfRange_Memory() in D:\a\1\s\src\System.Runtime\tests\System\StringTests.netcoreapp.cs:line 1560
```

Reply:
Looks like that last-minute [optimization](https://github.com/grant-d/coreclr/commit/ee5f321368daa6fe64226a3dd287285eb97b43bd#diff-cc88fbaa6077c440a3e08374eb41c5ea) broke the tests. Trying a local revert to see if it fixes them.

Reply:
Confirmed. All tests pass locally if I remove those `+ 1` clauses

Reply:
See CoreCLR PR that reverts this optimization: 
https://github.com/dotnet/coreclr/pull/23759